### ISSUE TITLE
Fix locking in TaskManager

### ISSUE THREAD

**Issue by ksavelev** [fluentscheduler/FluentScheduler#63]
Hi guys,

Why no locks around marked rows below?

```c#
class TaskManager
{
        private static void Schedule()
        {
            _timer.Stop();

            // not synchronized with AddTask and few other places where _tasks list can be modified
            // this can potentially end up with InvalidOperationException(Collection was modified; enumeration operation may not execute)
            var firstTask = _tasks.FirstOrDefault();

            // ...

                    // ok
                    lock (_tasksLock)
                    {
                        _tasks.Remove(firstTask);
                    }

            // ...

                // not synchronized with AddTask and few other places where _tasks list can be modified
                _tasks.Sort((x, y) => DateTime.Compare(x.NextRun, y.NextRun));
                Schedule();
                return;
            }
}
```

It is not expected to add new tasks after `_timer` started?

**Comment by tallesl** [comment]
Hey there.

Thanks for noticing. Yeah, I have picked that too.

I have planned a released closing #35 and also refactoring the TaskManager.
There are other issues as well (such as the timer not being disposed) not to mention the overall complexity of the class itself (I'm considering breaking it in internal classes).

I have been somewhat busy in this last weeks, but I hope it won't be long until that. In the meanwhile I'll ask you for a little patience.

(By the way, I don't know if you noticed, but the project was kinda abandoned from Sep 2014 to Sep 2015, I'm still taking the dust off around here...)

**Comment by ksavelev** [comment]
No I didn't know that project was abandoned. I've just started at new workplace, seen that people use this library and was curious how it is implemented, so glanced over the code.

And thanks for your effort BTW.
